deeppavlov/core/commands/train.py

```diff
@@ -278,237 +279,253 @@ def _train_batches(model: NNModel, iterator: DataLearningIterator, train_config:
+        def improved(score, best):
+            return score > best
 
-            if epochs >= train_config['epochs'] > 0:
-                break
-    except KeyboardInterrupt:
-        log.info('Stopped training')
+        train_monitoring['best'] = float('-inf')
+    elif train_config['metric_optimization'] == 'minimize':
+        def improved(score, best):
+            return score < best
+
+        train_monitoring['best'] = float('inf')
+    else:
+        raise ConfigError('metric_optimization has to be one of {}'.format(['maximize', 'minimize']))
+    train_monitoring['improved'] = improved
+
+    if train_config['tensorboard_log_dir'] is not None:
+        tb_log_dir = expand_path(train_config['tensorboard_log_dir'])
 
-    if not saved:
+        train_monitoring['tb_train_writer'] = tf.summary.FileWriter(str(tb_log_dir / 'train_log'))
+        train_monitoring['tb_valid_writer'] = tf.summary.FileWriter(str(tb_log_dir / 'valid_log'))
+    return train_monitoring
+
+
+def _clear_after_log(train_monitoring):
+    train_monitoring['losses'].clear()
+    train_monitoring['train_y_predicted'].clear()
+    train_monitoring['train_y_true'].clear()
+    return train_monitoring
+
+
+def _log(
+        model,
+        tag,
+        last_run,
+        train_monitoring,
+):
+    epochs = train_monitoring['epochs']
+
+    metrics = [
+        (s, f(train_monitoring['train_y_true'], train_monitoring['train_y_predicted']))
+        for s, f in train_monitoring['train_metrics_functions']
+    ]
+    report = dict(
+        epochs_done=epochs,
+        batches_seen=train_monitoring['batches_seen'],
+        train_examples_seen=train_monitoring['examples_seen'],
+        metrics=prettify_metrics(metrics),
+        time_spent=str(datetime.timedelta(seconds=round(time.time() - train_monitoring['start_time'] + 0.5)))
+    )
+
+    if train_monitoring['show_examples']:
+        try:
+            report['examples'] = [
+                dict(
+                    x=x_item,
+                    y_predicted=y_predicted_item,
+                    y_true=y_true_item
+                )
+                for x_item, y_predicted_item, y_true_item in zip(
+                    last_run['x'],
+                    last_run['y_predicted'],
+                    last_run['y_true'],
+                )
+            ]
+        except NameError:
+            log.warning('Could not log examples')
+
+    losses = train_monitoring['losses']
+    if train_monitoring['losses']:
+        report['loss'] = sum(losses) / len(losses)
+
+    tb_train_writer = train_monitoring['tb_train_writer']
+    if train_monitoring['tb_train_writer'] is not None:
+        for name, score in metrics:
+            metric_sum = tf.Summary(value=[tf.Summary.Value(tag='{}/'.format(tag) + name,
+                                                            simple_value=score), ])
+            tb_train_writer.add_summary(metric_sum, epochs)
+
+        if losses:
```

Comment:
this line should be changed to:
`if 'loss' in report:`
like it was done in PR: https://github.com/deepmipt/DeepPavlov/pull/450